tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,279 @@
+    $host_vf{bdf} = $host_vfs[int(rand(7))];  #pick up randomly
+    $host_vf{id} = get_device_id($host_vf{bdf});
+    record_info("Detach VF from host", "$host_vf{id}");
+    assert_script_run "virsh nodedev-detach $host_vf{id}";
+
+    #plugin vf to vm via virsh
+    record_info("Pluging vf to vm via libvirt", "guest=$guest \nvf=$host_vf{id}");
+    my %vf_vm = plugin_device_kvm($guest, $host_vf{id});
+    record_info("VF plugg'ed in", "bdf=$vf_vm{bdf}, mac=$vf_vm{mac}");
+  
+    #check the networking of the plugged interface
+    my $gateway = script_output "ip r s | grep 'default via' | cut -d ' ' -f3";
+    set_var("SRIOV_PCI_PASSSHTROUGH", 1);   #to differenciate virtual network tests
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm{mac}, net => 'br123');   #use br123 as ssh connection
+  
+    #reboot
+    record_info("VM rebooting", "$guest");
+    script_run "ssh root\@$guest 'reboot'";
+    my $time = 0;
+    while($time < 120 ) {
+      last if ! script_run "ssh root\@$guest 'echo Hello!'";
+      $time += 10;
+      sleep 10;
+    }
+  
+    #check again the networking inside vm
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm{mac}, net => 'br123');
+    set_var("SRIOV_PCI_PASSSHTROUGH", 0);   #turn off to avoid affecting other tests
+  
+    #detach vf from vm via virsh
+    record_info("Unplug VF from vm", "vm=$guest \nmac=$vf_vm{mac}"); 
+    if (unplug_vf_from_vm($guest, $vf_vm{mac})) {
+      $self->{test_results}->{$guest}->{"Error: fail to detach SR-IOV vf network device from $guest!"}->{status} = 'FAILED'; 
+    }
+  
+    #reattach vf to host
+    assert_script_run("virsh nodedev-reattach $host_vf{id}", 60);
+    script_run "lspci | grep Ethernet";
+    record_info("Reattach VF to host", "vm=$guest \nvf=$host_vf{id}");
+  
+    #check host and guest to make sure they work well
+    if (check_host_and_guest($guest, %host_vf)) {   #have not figured the parameters out???
+      $self->{test_results}->{$guest}->{"Error: host or guest $guest don't work well after vf, $host_vf{bdf}, is detached!"}->{status} = 'FAILED';
+    }
+
+  } #end each
+
+}
+
+#set up guest test environment
+sub prepare_guest() {
+  my $vm = shift;
+
+  #passwordless access to guest.
+  save_guest_ip($vm, name => "br123");     #get the guest ip via key words in 'virsh domiflist'
+  #if (! script_run("ssh root\@$guest 'ls /root/.ssh/authorized_keys'", 20)) {
+    exec_and_insert_password("ssh-copy-id -o StrictHostKeyChecking=no -i ~/.ssh/id_rsa.pub -f root\@$vm");
+  #}
+  save_screenshot;
+
+  #install required packages for test
+  zypper_call '-t in pciutils';  #to run 'lspci' command
+
+}
+
+#get the BDF the PF device on host
+sub get_sriov_devices () {
+
+  #get the BDF of the ethernet devices with SR-IOV
+  my $nic_devices = script_output "lspci | grep Ethernet | grep -v 'Virtual Function' | cut -d ' ' -f1" ;
+  my @nic_devices = split ("\n", $nic_devices);
+  print "julies: \@nic_devices=@nic_devices\n";
```

Comment:
Is there the debug info, right?

Reply:
Yes, well recognized by searching my name :-) it was missed when i removed all the debug info just before submitting pr